src/Core/Field.h

```diff
@@ -1029,3 +1029,92 @@ struct fmt::formatter<DB::Field>
         return fmt::format_to(ctx.out(), "{}", toString(x));
     }
 };
+
+namespace std
+{
+    template <>
+    struct hash<DB::Field>
+    {
+        size_t operator()(const DB::Field & field) const noexcept
+        {
+            using Which = DB::Field::Types::Which;
+            size_t type_hash = std::hash<UInt8>{}(field.getType());
+            switch (field.getType())
+            {
+            case Which::Null:
+                return 0;
+            case Which::Bool:
+                [[fallthrough]];
+            case Which::UInt64:
+                return type_hash ^ std::hash<UInt64>{}(field.get<UInt64>());
```

Comment:
`std::hash` is very bad for integers, see:
https://presentations.clickhouse.com/hash_tables/#8

Xor of hashes is very bad for hash combining, see:
https://presentations.clickhouse.com/hash_tables/#11

Reply:
Reimplementing with intHash and `hash(x) ^ y`.

Reply:
After implementing it I've decided to instead fix `Field::operator<()`. I might keep the implementation in a separate file in case we want to use it in the future (although in a different file, to avoid pushing even more deps to Field.h